### ISSUE TITLE
libtorch windows VS2019 projects pendeling on return

### ISSUE THREAD

**Issue by VitaMusic** [pytorch/pytorch#30322]
## 🐛 Bug

I use downloaded from site a stable version 1.3.1 of libtorch in debug and release modes
And I'm testing a simple project on mnist data from pytorch site.

When project is executed and the main have to exit after return 0; it does not happens, neigther debug nor release modes. the program is pendeling and the process have to be killed.

## To Reproduce

cpu only, libtorch debug or release, VS2019 with last toolchain, cmake for making VS project

`#include <torch/torch.h>

// Define a new Module.
struct Net : torch::nn::Module {
	Net() {
		// Construct and register two Linear submodules.
		fc1 = register_module("fc1", torch::nn::Linear(784, 64));
		fc2 = register_module("fc2", torch::nn::Linear(64, 32));
		fc3 = register_module("fc3", torch::nn::Linear(32, 10));
	}

	// Implement the Net's algorithm.
	torch::Tensor forward(torch::Tensor x) {
		// Use one of many tensor manipulation functions.
		x = torch::relu(fc1->forward(x.reshape({ x.size(0), 784 })));
		x = torch::dropout(x, /*p=*/0.5, /*train=*/is_training());
		x = torch::relu(fc2->forward(x));
		x = torch::log_softmax(fc3->forward(x), /*dim=*/1);
		return x;
	}

	// Use one of many "standard library" modules.
	torch::nn::Linear fc1{ nullptr }, fc2{ nullptr }, fc3{ nullptr };
};

int main() {
	// Create a new Net.
	auto net = std::make_shared<Net>();

	// Create a multi-threaded data loader for the MNIST dataset.
	auto data_loader = torch::data::make_data_loader(
		torch::data::datasets::MNIST("./data").map(
			torch::data::transforms::Stack<>()),
		/*batch_size=*/64);

	// Instantiate an SGD optimization algorithm to update our Net's parameters.
	torch::optim::SGD optimizer(net->parameters(), /*lr=*/0.01);

	for (size_t epoch = 1; epoch <= 1; ++epoch) {
		size_t batch_index = 0;
		// Iterate the data loader to yield batches from the dataset.
		for (auto& batch : *data_loader) {
			// Reset gradients.
			optimizer.zero_grad();
			// Execute the model on the input data.
			torch::Tensor prediction = net->forward(batch.data);
			// Compute a loss value to judge the prediction of our model.
			torch::Tensor loss = torch::nll_loss(prediction, batch.target);
			// Compute gradients of the loss w.r.t. the parameters of our model.
			loss.backward();
			// Update the parameters based on the calculated gradients.
			optimizer.step();
			// Output the loss and checkpoint every 100 batches.
			if (++batch_index % 100 == 0) {
				std::cout << "Epoch: " << epoch << " | Batch: " << batch_index
					<< " | Loss: " << loss.item<float>() << std::endl;
				// Serialize your model periodically as a checkpoint.
				torch::save(net, "net.pt");
			}
		}
	}
	std::cout << "END LEARNING";
	return EXIT_SUCCESS;		
}`

1. Create CMake project with an example code
2. Configure and generate VS2019 project
3. Run the project in VS in debug or release mode (download of mnist data is needed)

## Expected behavior

exit the main program when finishing program execution.

## Environment

 - PyTorch Version (e.g., 1.0): non conserned, libtorch from the site, v1.3.1
 - OS (e.g., Linux): Windows 7 x64 Pro
 - How you installed PyTorch (`conda`, `pip`, source): download from the site
 - Build command you used (if compiling from source): not concerned
 - Python version: not concerned
 - CUDA/cuDNN version: not concerned
 - GPU models and configuration: not concerned
 - Any other relevant information: last CMake and last VS2019 installed

cc @ezyang @peterjc123

**Comment by peterjc123** [comment]
Do you know where it gets stuck? Actually, it is somehow related to https://github.com/pytorch/pytorch/issues/29383 that only reproducible on Windows 7. Could you please post the stacktrace to help us dig into it?

**Comment by VitaMusic** [comment]
I tried the case #29383 I can reproduce it. For my case I do not know how to do the stacktrace in my case, if you explain me I'll do. The problem is after return 0; line sur the call stack is empty in VS.
I noted also that libtorch is compiled with vs_runtime 140, could this be a problem ?
I do not know how to recompile it from source, the compilation does not work for me.

**Comment by peterjc123** [comment]
@VitaMusic You said it didn't return. So just click the pause button during the debug session of VS, and the highlighted line will be the one of question. As for stacktrace, it is also called call stack, looks like the following:
```
functionA() line 23 a.cpp
functionB() line 45 b.cpp
...

**Comment by VitaMusic** [comment]
here is the message I have: 
Your app has entered a break state, but there is no code to show because all threads were executing external code (typically system or framework code).

![image](https://user-images.githubusercontent.com/12452501/69563404-52bd7f00-0fb1-11ea-9d5e-c7a36db1a513.png)


output:
'libtorch.exe' (Win32): Loaded 'D:\PROJECTS\libtorch\neuroevolution\x64\Debug\libtorch.exe'. Symbols loaded.
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\ntdll.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\kernel32.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\KernelBase.dll'. 
'libtorch.exe' (Win32): Loaded 'D:\PROJECTS\libtorch\neuroevolution\x64\Debug\c10.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\msvcp140d.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\vcruntime140d.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\ucrtbased.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\api-ms-win-core-localization-l1-2-0.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\api-ms-win-core-processthreads-l1-1-1.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\api-ms-win-core-file-l1-2-0.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\api-ms-win-core-timezone-l1-1-0.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\api-ms-win-core-file-l2-1-0.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\api-ms-win-core-synch-l1-2-0.dll'. 
'libtorch.exe' (Win32): Loaded 'D:\PROJECTS\libtorch\neuroevolution\x64\Debug\torch.dll'. 
'libtorch.exe' (Win32): Loaded 'C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2019.5.281\windows\redist\intel64_win\compiler\libiomp5md.dll'. Symbols loaded.
'libtorch.exe' (Win32): Loaded 'C:\Windows\System32\vcruntime140_1d.dll'. 
The thread 0x2d2c has exited with code 0 (0x0).
The thread 0x1c7c has exited with code 0 (0x0).
The thread 0x2dd4 has exited with code 0 (0x0).
The thread 0x19d4 has exited with code 0 (0x0).
The thread 0x8fc has exited with code 0 (0x0).
The thread 0x1cac has exited with code 0 (0x0).
The thread 0x1ca0 has exited with code 0 (0x0).

**Comment by peterjc123** [comment]
I see, what about using breakpoints and then step into?
BTW, Why is it using the `libiomp5md.dll` from your OS? Could you please copy that one from libtorch?

**Comment by VitaMusic** [comment]
> Why is it using the `libiomp5md.dll` from your OS? Could you please copy that one from libtorch?

I have newer version, I tried the both, no difference 

----------------------------
call stack : 
>	libtorch.exe!std::unique_ptr<torch::data::StatelessDataLoader<torch::data::datasets::MapDataset<torch::data::datasets::MNIST,torch::data::transforms::Stack<torch::data::Example<at::Tensor,at::Tensor>>>,torch::data::samplers::RandomSampler>,std::default_delete<torch::data::StatelessDataLoader<torch::data::datasets::MapDataset<torch::data::datasets::MNIST,torch::data::transforms::Stack<torch::data::Example<at::Tensor,at::Tensor>>>,torch::data::samplers::RandomSampler>>>::~unique_ptr<torch::data::StatelessDataLoader<torch::data::datasets::MapDataset<torch::data::datasets::MNIST,torch::data::transforms::Stack<torch::data::Example<at::Tensor,at::Tensor>>>,torch::data::samplers::RandomSampler>,std::default_delete<torch::data::StatelessDataLoader<torch::data::datasets::MapDataset<torch::data::datasets::MNIST,torch::data::transforms::Stack<torch::data::Example<at::Tensor,at::Tensor>>>,torch::data::samplers::RandomSampler>>>() Line 1873	C++

![image](https://user-images.githubusercontent.com/12452501/69636488-fd3bad80-1056-11ea-913b-7336427be639.png)

![image](https://user-images.githubusercontent.com/12452501/69636932-f6fa0100-1057-11ea-9105-287ba2a142c8.png)

vs runtime block the things

**Comment by peterjc123** [comment]
There must be some threads that are not terminated.

**Comment by VitaMusic** [comment]
> There must be some threads that are not terminated.

yes

**Comment by VitaMusic** [comment]
I had the similar things with threads when I used cilk++, I had to terminated the threads explicitly in the code.

**Comment by peterjc123** [comment]
No, usually we don't have to do this. There must be a deadlock or something. Could you please dig into it a bit more and find where it gets stuck?

**Comment by VitaMusic** [comment]
Give me more detailed instructions on how to do please.

ср, 27 Ноя 2019 г., 6:42 peterjc123 <notifications@github.com>:

**Comment by iremyux** [comment]
Hi @VitaMusic , using Libtorch 2.0.1, I was unable to reproduce this issue. If you are still experiencing the same problem, can you please give me more details? If you could provide me with a dump file, it would greatly assist me in addressing your problem.

**Comment by VitaMusic** [comment]
I don't work on this project anymore.

вт, 30 мая 2023 г., 5:08 AM Irem Yuksel ***@***.***>:

**Comment by iremyux** [comment]
Based on the current information, I have decided to close this issue as further investigation may not be necessary at this time.
However, please don't hesitate to open a new issue if you encounter similar or different problems in the future.
